### ISSUE TITLE
Rule proposal: `no-unnecessary-rename`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1313]
Sometimes, when we import a module or destructuring an object, the property name are not available, we have to rename it. But during refactor, the property name is available again, we may want to avoid rename.

```diff
- const foo = 1;
const {foo: anotherFoo} = bar;
```

Since we removed `foo`, I want to use `foo` instead of `anotherFoo` now.

## Fail

```js
const {foo: anotherFoo} = bar;
```

```js
import {fromPairs as lodashFromPairs} from 'lodash-es';
```

## Pass

```js
const foo = 1;
const {foo: anotherFoo} = bar;
```

```js
import {fromPairs as lodashFromPairs} from 'lodash-es';

function fromPairs() {
  // Do something different from `lodash.fromPairs`
}
```

**Comment by papb** [comment]
> Or
> 
> ```diff
> const foo = 1;
> + if (bar) {
> 	const {foo: anotherFoo} = bar; // `foo` is available in the new scope now, and I don't need access to the outer `foo`.
> + }
> ```

Detecting this would probably conflict with `no-shadow`, right? I think only the first part of what you proposed should be done.

**Comment by fisker** [comment]
You are right. For people like me may don't care about `no-shadow`, maybe we can put this check behind under an option.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by fregante** [comment]
Your examples are good but conflicts aren’t the only reason to rename a variable, they’re more likely the exception. 

```js
const {textContent: age} = ageInputElement;
```

I think you’re suggesting I should name the constant `textContent` instead of `age` and I’m sure we agree that’s a bad idea.

**Comment by sindresorhus** [comment]
i think we should close this as not worth adding.

The objective case is already covered by ESLint's `no-useless-rename`, where the imported or destructured name is renamed to itself. This proposal goes further and treats an available original name as evidence that an alias is unnecessary, but that is not generally true.

Aliases are often intentional semantic names:

```js
const {textContent: age} = ageInputElement;
const {arguments: argumentNodes} = node;
const {parse: parseRegExp} = regjsparser;
```

Those are better than the original property names. The rule would either report too many good aliases, or need allowlists and heuristics to guess intent.

I also do not think the shadowing option is worth it. It conflicts with `no-shadow`, and an autofix could change binding resolution or introduce TDZ behavior. Without autofix, the rule becomes mostly subjective noise.

So my vote is to rely on `no-useless-rename`, `object-shorthand`, `no-shadow`, and `prevent-abbreviations` for the objective parts, and leave stale aliases after refactors to editor or code serch workflows.
